File: app/internal/publisher/template/go/welcome.go
package wsd

import (
	"bufio"
	"fmt"
	"os"
	"strings"
)

type WelcomeDialog struct {
	engine      *LicenseEngine
	productName string
}

func NewWelcomeDialog(engine *LicenseEngine, productName string) *WelcomeDialog {
	return &WelcomeDialog{
		engine:      engine,
		productName: productName,
	}
}

func (d *WelcomeDialog) IsOnboardingComplete() bool {
	return d.engine.cache.IsOnboardingComplete()
}

func (d *WelcomeDialog) Show() map[string]interface{} {
	if d.IsOnboardingComplete() {
		fmt.Println("Onboarding already completed.")
		return map[string]interface{}{"skipped": true, "message": "Onboarding already completed"}
	}
	if !d.engine.config.Trial.Enabled {
		fmt.Println("Trial onboarding is not enabled.")
		return map[string]interface{}{"skipped": true, "message": "Trial onboarding is not enabled"}
	}
	reader := bufio.NewReader(os.Stdin)
	fmt.Println(strings.Repeat("=", 60))
	title := d.productName
	if title == "" {
		title = d.engine.config.Product.Name
	}
	if title == "" {
		title = "Software"
	}
	fmt.Printf("Welcome to %s\n", title)
	fmt.Println("Complete your registration to start the trial")
	fmt.Println(strings.Repeat("=", 60))
	fmt.Print("Name *: ")
	name, _ := reader.ReadString('\n')
	name = strings.TrimSpace(name)
	if name == "" {
		fmt.Println("Name is required.")
		return map[string]interface{}{"skipped": true, "error": "Name is required"}
	}
	fmt.Print("Email *: ")
	email, _ := reader.ReadString('\n')
	email = strings.TrimSpace(email)
	if email == "" || !strings.Contains(email, "@") {
		fmt.Println("Valid email is required.")
		return map[string]interface{}{"skipped": true, "error": "Valid email is required"}
	}
	fmt.Print("Mobile Number *: ")
	mobile, _ := reader.ReadString('\n')
	mobile = strings.TrimSpace(mobile)
	if mobile == "" || len(mobile) < 4 {
		fmt.Println("Valid mobile number is required.")
		return map[string]interface{}{"skipped": true, "error": "Valid mobile number is required"}
	}
	fmt.Print("Company (optional): ")
	company, _ := reader.ReadString('\n')
	company = strings.TrimSpace(company)
	fmt.Println()
	fmt.Println("Sending OTP to your email...")
	otpResult, err := d.engine.client.request("auth/otp/send", map[string]interface{}{
		"email": email,
	})
	if err != nil || otpResult == nil || !otpResult["success"].(bool) {
		fmt.Println("Failed to send OTP. Continuing directly...")
	}
	fmt.Print("Enter OTP (or press Enter to skip): ")
	otp, _ := reader.ReadString('\n')
	otp = strings.TrimSpace(otp)
	if otp != "" {
		verifyResult, err := d.engine.client.request("auth/otp/verify", map[string]interface{}{
			"email": email,
			"otp":   otp,
		})
		if err == nil && verifyResult != nil {
			if success, ok := verifyResult["success"].(bool); ok && success {
				fmt.Println("OTP verified!")
			} else {
				fmt.Println("OTP verification failed, but continuing...")
			}
		}
	}
	fmt.Println("Activating trial...")
	hardwareID := d.engine.GetHardwareID()
	customerData := map[string]interface{}{
		"mobile":      mobile,
		"hardware_id": hardwareID,
	}
	if company != "" {
		customerData["company_name"] = company
	}
	_, err = d.engine.client.request("customer/register", map[string]interface{}{
		"name":        name,
		"email":       email,
		"mobile":      mobile,
		"hardware_id": hardwareID,
	})
	if err != nil {
		fmt.Printf("Warning: Customer registration issue: %s\n", err.Error())
	}
	trialResult, err := d.engine.StartTrial(email, name, customerData)
	if err != nil {
		fmt.Printf("Trial activation failed: %s\n", err.Error())
		return map[string]interface{}{
			"skipped":  false,
			"error":    err.Error(),
			"name":     name,
			"email":    email,
		}
	}
	if success, ok := trialResult["success"].(bool); ok && success {
		d.engine.cache.SetOnboardingComplete()
		fmt.Println("Trial activated! You can now use the software.")
		return map[string]interface{}{
			"name":               name,
			"email":              email,
			"hardware_id":        hardwareID,
			"onboarding_complete": true,
		}
	}
	fmt.Println("Trial activation had issues, but setup completed.")
	d.engine.cache.SetOnboardingComplete()
	return map[string]interface{}{
		"name":               name,
		"email":              email,
		"hardware_id":        hardwareID,
		"onboarding_complete": true,
	}
}

File: app/internal/publisher/template/rust/src/welcome.rs
use std::io::{self, BufRead, Write};

use crate::license_engine::LicenseEngine;

pub struct WelcomeDialog;

impl WelcomeDialog {
    pub fn show(engine: &mut LicenseEngine) -> Result<Option<WelcomeResult>, String> {
        if engine.is_onboarding_complete() {
            let _ = writeln!(io::stdout(), "Onboarding already completed.");
            return Ok(None);
        }
        if !engine.config.trial.enabled {
            let _ = writeln!(io::stdout(), "Trial onboarding is not enabled.");
            return Ok(None);
        }

        let stdin = io::stdin();
        let mut stdout = io::stdout();

        let _ = writeln!(stdout, "{}", "=".repeat(60));
        let title = if engine.product_name.is_empty() {
            engine.config.product.name.clone()
        } else {
            engine.product_name.clone()
        };
        let title = if title.is_empty() { "Software".to_string() } else { title };
        let _ = writeln!(stdout, "Welcome to {}", title);
        let _ = writeln!(stdout, "Complete your registration to start the trial");
        let _ = writeln!(stdout, "{}", "=".repeat(60));

        let _ = write!(stdout, "Name *: ");
        let _ = stdout.flush();
        let mut name = String::new();
        stdin.lock().read_line(&mut name).map_err(|e| e.to_string())?;
        let name = name.trim().to_string();
        if name.is_empty() {
            let _ = writeln!(stdout, "Name is required.");
            return Ok(None);
        }

        let _ = write!(stdout, "Email *: ");
        let _ = stdout.flush();
        let mut email = String::new();
        stdin.lock().read_line(&mut email).map_err(|e| e.to_string())?;
        let email = email.trim().to_string();
        if email.is_empty() || !email.contains('@') {
            let _ = writeln!(stdout, "Valid email is required.");
            return Ok(None);
        }

        let _ = write!(stdout, "Mobile Number *: ");
        let _ = stdout.flush();
        let mut mobile = String::new();
        stdin.lock().read_line(&mut mobile).map_err(|e| e.to_string())?;
        let mobile = mobile.trim().to_string();
        if mobile.is_empty() || mobile.len() < 4 {
            let _ = writeln!(stdout, "Valid mobile number is required.");
            return Ok(None);
        }

        let _ = write!(stdout, "Company (optional): ");
        let _ = stdout.flush();
        let mut company = String::new();
        stdin.lock().read_line(&mut company).map_err(|e| e.to_string())?;
        let company = company.trim().to_string();

        let _ = writeln!(stdout);
        let _ = writeln!(stdout, "Sending OTP to your email...");
        let _ = stdout.flush();

        let otp_payload = serde_json::json!({"email": email});
        let _ = engine.client.request("auth/otp/send", otp_payload);

        let _ = write!(stdout, "Enter OTP (or press Enter to skip): ");
        let _ = stdout.flush();
        let mut otp = String::new();
        stdin.lock().read_line(&mut otp).map_err(|e| e.to_string())?;
        let otp = otp.trim().to_string();

        if !otp.is_empty() {
            let verify_payload = serde_json::json!({"email": email, "otp": otp});
            match engine.client.request("auth/otp/verify", verify_payload) {
                Ok(res) => {
                    if res.get("success").and_then(|v| v.as_bool()).unwrap_or(false) {
                        let _ = writeln!(stdout, "OTP verified!");
                    }
                }
                Err(_) => {
                    let _ = writeln!(stdout, "OTP verification failed, but continuing...");
                }
            }
        }

        let _ = writeln!(stdout, "Activating trial...");
        let _ = stdout.flush();

        let hardware_id = engine.get_hardware_id();

        let register_payload = serde_json::json!({
            "name": name,
            "email": email,
            "mobile": mobile,
            "hardware_id": hardware_id,
        });
        let _ = engine.client.request("customer/register", register_payload);

        let mut customer_data = serde_json::json!({
            "mobile": mobile,
            "hardware_id": hardware_id,
        });
        if !company.is_empty() {
            customer_data["company_name"] = serde_json::json!(company);
        }

        match engine.start_trial(&email, &name, Some(customer_data)) {
            Ok(result) => {
                if result.get("success").and_then(|v| v.as_bool()).unwrap_or(false) {
                    engine.set_onboarding_complete();
                    let _ = writeln!(stdout, "Trial activated! You can now use the software.");
                    Ok(Some(WelcomeResult {
                        name,
                        email,
                        hardware_id,
                        onboarding_complete: true,
                    }))
                } else {
                    let _ = writeln!(stdout, "Trial activation had issues, but setup completed.");
                    engine.set_onboarding_complete();
                    Ok(Some(WelcomeResult {
                        name,
                        email,
                        hardware_id,
                        onboarding_complete: true,
                    }))
                }
            }
            Err(e) => {
                let _ = writeln!(stdout, "Trial activation failed: {}", e);
                Ok(Some(WelcomeResult {
                    name,
                    email,
                    hardware_id,
                    onboarding_complete: false,
                }))
            }
        }
    }
}

pub struct WelcomeResult {
    pub name: String,
    pub email: String,
    pub hardware_id: String,
    pub onboarding_complete: bool,
}
